src/modules/graphics/vulkan/Shader.h

```diff
@@ -0,0 +1,153 @@
+	virtual ~Shader();
+
+	bool loadVolatile() override;
+	void unloadVolatile() override;
+
+	VkPipeline getComputePipeline() const;
+
+	const std::vector<VkPipelineShaderStageCreateInfo> &getShaderStages() const;
+
+	const VkPipelineLayout getGraphicsPipelineLayout() const;
+
+	void newFrame(uint32_t frameIndex);
+
+	void cmdPushDescriptorSets(VkCommandBuffer, VkPipelineBindPoint);
+
+	void attach() override;
+
+	ptrdiff_t getHandle() const { return 0; }
+
+	std::string getWarnings() const override { return ""; }
+
+	int getVertexAttributeIndex(const std::string &name) override;
+
+	const UniformInfo *getUniformInfo(const std::string &name) const override;
+	const UniformInfo *getUniformInfo(BuiltinUniform builtin) const override;
+
+	void updateUniform(const UniformInfo *info, int count) override;
+
+	void sendTextures(const UniformInfo *info, graphics::Texture **textures, int count) override;
+	void sendBuffers(const UniformInfo *info, love::graphics::Buffer **buffers, int count) override;
+
+	bool hasUniform(const std::string &name) const override;
+
+	void setVideoTextures(graphics::Texture *ytexture, graphics::Texture *cbtexture, graphics::Texture *crtexture) override;
+
+	void setMainTex(graphics::Texture *texture);
+
+private:
+	void calculateUniformBufferSizeAligned();
+	void compileShaders();
+	void createDescriptorSetLayout();
+	void createPipelineLayout();
+	void createDescriptorPoolSizes();
+	void createStreamBuffers();
+	void buildLocalUniforms(
+		spirv_cross::Compiler &comp, 
+		const spirv_cross::SPIRType &type, 
+		size_t baseoff, 
+		const std::string &basename);
+	void initDescriptorSet();
+	void updateUniform(const UniformInfo* info, int count, bool internal);
+
+	VkDescriptorSet allocateDescriptorSet();
+
+	VkDeviceSize uniformBufferSizeAligned;
+
+	VkPipeline computePipeline;
+
+	VkDescriptorSetLayout descriptorSetLayout;
+	VkPipelineLayout pipelineLayout;
+	std::vector<VkDescriptorPoolSize> descriptorPoolSizes;
+
+	// we don't know how much memory we need per frame for the uniform buffer descriptors
+	// we keep a vector of stream buffers per frame in flight
+	// that gets dynamically increased if more memory is needed
+	std::vector<std::vector<StreamBuffer*>> streamBuffers;
+	std::vector<VkDescriptorPool> descriptorPools;
+	std::queue<VkDescriptorSet> freeDescriptorSets;
+	std::vector<std::vector<VkDescriptorSet>> descriptorSetsVector;
+
+	std::vector<VkPipelineShaderStageCreateInfo> shaderStages;
+	std::vector<VkShaderModule> shaderModules;
+
+	Graphics *vgfx = nullptr;
+	VkDevice device;
+
+	bool isCompute = false;
+
+	std::unordered_map<std::string, graphics::Shader::UniformInfo> uniformInfos;
+	UniformInfo *builtinUniformInfo[BUILTIN_MAX_ENUM];
+
+	std::unique_ptr<StreamBuffer> uniformBufferObjectBuffer;
+	std::vector<uint8> localUniformData;
+	std::vector<uint8> localUniformStagingData;
+	uint32_t uniformLocation;
+	std::optional<size_t> builtinUniformDataOffset;
```

Comment:
This could use love's `OptionalBool` type (from `common/Optional.h`), std::optional needs a pretty new C++ standard library on the user's system so we might want to stick with love's existing hand-made Optional stuff instead, for now.